daemon/status.go

```diff
@@ -85,90 +81,275 @@ func (d *Daemon) getNodeStatus() *models.ClusterStatus {
 	return &clusterStatus
 }
 
-func (d *Daemon) collectStatus() {
-	for {
-		response := d.getStatus()
-
-		d.statusCollectMutex.Lock()
-		d.statusResponse = response
-		d.statusResponseTimestamp = time.Now()
-		d.statusCollectMutex.Unlock()
-
-		time.Sleep(collectStatusInterval)
-	}
-}
-
-func (d *Daemon) startStatusCollector() {
-	go d.collectStatus()
-}
-
 func (h *getHealthz) Handle(params GetHealthzParams) middleware.Responder {
-	d := h.daemon
-	d.statusCollectMutex.RLock()
-	sr := d.statusResponse
-	timestamp := d.statusResponseTimestamp
-	d.statusCollectMutex.RUnlock()
-
-	if time.Since(timestamp) > staleTimeout {
-		sr.Cilium = &models.Status{
-			State: models.StatusStateWarning,
-			Msg:   fmt.Sprintf("Stale status data (since %v)", timestamp),
-		}
-	}
+	sr := h.daemon.getStatus()
 
 	return NewGetHealthzOK().WithPayload(&sr)
 }
 
 func (d *Daemon) getStatus() models.StatusResponse {
-	sr := models.StatusResponse{
-		Controllers: controller.GetGlobalStatus(),
-	}
-
-	checkLocks(d)
-
-	if info, err := kvstore.Client().Status(); err != nil {
-		sr.Kvstore = &models.Status{State: models.StatusStateFailure, Msg: fmt.Sprintf("Err: %s - %s", err, info)}
-	} else {
-		sr.Kvstore = &models.Status{State: models.StatusStateOk, Msg: info}
+	staleProbes := d.statusCollector.GetStaleProbes()
+	stale := make(map[string]strfmt.DateTime, len(staleProbes))
+	for probe, startTime := range staleProbes {
+		stale[probe] = strfmt.DateTime(startTime)
 	}
 
-	sr.ContainerRuntime = workloads.Status()
+	d.statusCollectMutex.RLock()
+	defer d.statusCollectMutex.RUnlock()
 
-	sr.Kubernetes = d.getK8sStatus()
+	// TODO(brb) we should do a deep copy instead, as statusResponse contains some references
```

Comment:
And why can't we do a deep copy now?

Reply:
I added a dependency for doing a deep copy in this particular case, which is IMHO a bit an overkill. Alternatively, I could add a method for the deep copy manually, or just do `Unmarshal(Marshal(...))`.